Add derivative tally based k_eff search support for material perturbations

pranavkantgaur opened the pull request:
<!--
If you are a first-time contributor to OpenMC, please have a look at our
contributing guidelines:
https://github.com/openmc-dev/openmc/blob/develop/CONTRIBUTING.md
-->

# Description

This PR extends the recently added Model.keff_search method for material input parameters by leveraging derivative tallies. It adds least-squares curve fitting over a combination of K_eff evaluations and its derivatives as an alternative method to GRSecant's curve fitting. It is authored entirely using Copilot and reviewed by me. We have observed considerable reduction in the required number of MC evaluations for some cases (reproducible using `test_tally_deriv_keff_search.py` in the [attached zip](https://github.com/user-attachments/files/24490117/keff_search_derivatives.zip)). Results are consistent with the primary work by @smharper as reported [here](https://dspace.mit.edu/bitstream/handle/1721.1/106690/969775837-MIT.pdf).

It is an initial work on #980 and is related to: #3569, #2693. 



Results from the attached demo script are as:


#### TABLE 1: BORON CONCENTRATION SEARCH RESULTS

| Method                      | Final Sol (ppm) | MC Runs | Tot Batches | Time (s) | Converged |
|-----------------------------|-----------------|---------|-------------|----------|-----------|
| GRsecant (no deriv)         | 20.4            | 17      | 2282        | 55.48    | True      |
| Least Squares (with deriv)  | 30.0            | 9       | 1090        | 31.20    | True      |

##### Efficiency Gains (relative to GRsecant baseline)

**Least Squares (with deriv):**
- **MC runs:** +47.1% (9 vs 17)
- **Total batches:** +52.2% (1090 vs 2282)
- **Elapsed time:** +43.8% (31.20s vs 55.48s)

---

#### TABLE 2: FUEL DENSITY SEARCH RESULTS

| Method                      | Final Density (g/cm³) | MC Runs | Tot Batches | Time (s) | Converged |
|-----------------------------|-----------------------|---------|-------------|----------|-----------|
| GRsecant (no deriv)         | 7.007                 | 43      | 9627        | 228.95   | True      |
| Least Squares (with deriv)  | 6.978                 | 27      | 4783        | 137.50   | True      |

##### Efficiency Gains (relative to GRsecant baseline)

**Least Squares (with deriv):**
- **MC runs:** +37.2% (27 vs 43)
- **Total batches:** +50.3% (4783 vs 9627)
- **Elapsed time:** +39.9% (137.50s vs 228.95s)


# Checklist

- [x] I have performed a self-review of my own code
- [ ] I have run [clang-format](https://docs.openmc.org/en/latest/devguide/styleguide.html#automatic-formatting) (version 15) on any C++ source files (if applicable)
- [x] I have followed the [style guidelines](https://docs.openmc.org/en/latest/devguide/styleguide.html#python) for Python source files (if applicable)
- [ ] I have made corresponding changes to the documentation (if applicable)
- [x] I have added tests that prove my fix is effective or that my feature works (if applicable)
<!--
While tests will automatically be checked by CI, it is good practice to
ensure that they pass locally first. See instructions here:
https://docs.openmc.org/en/latest/devguide/tests.html
-->

pranavkantgaur committed:
add grsecant changes and test script with copilot

pranavkantgaur committed:
revert synthetic points and add derivative constraints copilot

pranavkantgaur committed:
add fixes for derivative scaling

pranavkantgaur committed:
add gradient descent

pranavkantgaur committed:
add geom search test, uncertainity weighing in GD

pranavkantgaur committed:
add flag for deriv uncertainity

pranavkantgaur committed:
add flag for off derv. uncertainity only but not derivatives

pranavkantgaur committed:
Merge branch 'openmc-dev:develop' into 980_k_search_tally_dervs

pranavkantgaur committed:
remove GD and add tests

pranavkantgaur committed:
remove  derivative feature flags

mvineetmenon committed:
Added unit tests for tally derivatives extension in keff_search

pranavkantgaur committed:
move tally obj. instantiation to model class

pranavkantgaur committed:
move compare script to examples dir

pranavkantgaur committed:
Merge branch 'openmc-dev:develop' into 980_k_search_tally_dervs

pranavkantgaur committed:
update test assertions, model creation and file rename

pranavkantgaur committed:
update comments

GuySten commented on examples/keff_search_derivatives/__init__.py:
The examples you added are nice but imo a better place for them should be in the [openmc-notebooks](https://github.com/openmc-dev/openmc-notebooks) repository.

pranavkantgaur replied:
Hi @GuySten, thanks for the comments.

An alternative to what @mvineetmenon suggested could be for us to refactor the gradient based least-squares logic out of the search_keff method as a standalone helper function and use that in a demo notebook, so that we can make the case for this approach to be included into the search_keff method.

In that direction, we did add a PR some time back in the openmc-notebooks repo: https://github.com/openmc-dev/openmc-notebooks/pull/31

Let us know if you recommend to maybe update that PR to say also add gradient-based least-squares approach in addition to the gradient-descent optimization (proposal of that PR) for k_eff search.

GuySten commented on examples/keff_search_derivatives/README.md:
See comment above.

GuySten commented on examples/keff_search_derivatives/test_tally_deriv_keff_search.py:
See comment above.

GuySten commented on tests/unit_tests/test_model.py:
I suggest moving the criticality search tests into test_criticality_search.py file.

pranavkantgaur replied:
We do have an existing criticality search test for 'model' class [test_keff_search](https://github.com/openmc-dev/openmc/blob/develop/tests/unit_tests/test_model.py#L946), do you suggest to move that out as well?

GuySten replied:
Yes, because it makes sense to me to group the criticality search tests together.

pranavkantgaur replied:
also tagging @paulromano to take his view on whether to move out [test_keff_search](https://github.com/openmc-dev/openmc/blob/develop/tests/unit_tests/test_model.py#L946) into a separate test_criticality_search.py as he added it [here](https://github.com/openmc-dev/openmc/pull/3569/changes#diff-9bebc58f92eb78e015c77d1c5fe99ae7081cc0bc1676cbcfda0907b9cc41eb5e) recently.

pranavkantgaur replied:
@GuySten I have moved out all criticality search tests into a separate test_criticality_search.py, could be rolled back later if required.

GuySten reviewed (changes requested):
This looks like a nice feature.
A better place for the examples should be in the openmc-notebooks repo though.

mvineetmenon commented:
@guysten

Thanks for the suggestion! This PR introduces a new feature enhancement to OpenMC by extending Model.keff_search to support material perturbations using derivative tallies (with least-squares fitting, uncertainty weighting, etc.).

Since this is a new capability being added in this PR, any accompanying notebook demonstrations may go to the separate [openmc-notebooks](https://github.com/openmc-dev/openmc-notebooks) repository rather than being included here. The notebooks in openmc-notebooks are intended as demonstrations and tutorials for existing features of released versions of OpenMC, not for showcasing in-development changes.

Once this PR is merged and the feature lands in a release, we can open a PR there to add a notebook example! That would be a great way to highlight the new functionality.

GuySten commented:
I thinks that the examples folder in openmc has alot less visibility than the [openmc-notebooks](https://github.com/openmc-dev/openmc-notebooks) repository. You've clearly put alot of work in the examples in this PR and I think it will be better to put them in a place that is more visible to users. 
I personally also like examples in the format of jupyter notebooks so you can more easily test the new feature.

What I meant was lets remove the example files in this PR and merge it without them.
And once it is merged you should open a PR in openmc-notebooks with the example files converted into notebooks.
That way more users will be exposed to this feature.

You can also add the examples to the description of this PR so people can run it while it still in development, if you want.

pranavkantgaur commented:
> I thinks that the examples folder in openmc has alot less visibility than the [openmc-notebooks](https://github.com/openmc-dev/openmc-notebooks) repository. You've clearly put alot of work in the examples in this PR and I think it will be better to put them in a place that is more visible to users. I personally also like examples in the format of jupyter notebooks so you can more easily test the new feature.
> 
> What I meant was lets remove the example files in this PR and merge it without them. And once it is merged you should open a PR in openmc-notebooks with the example files converted into notebooks. That way more users will be exposed to this feature.
> 
> You can also add the examples to the description of this PR so people can run it while it still in development, if you want.

Sure, I will take out the examples from this PR and add them as demo scripts in the description. Later maybe update the existing PR in openmc-notebooks with this example.

pranavkantgaur committed:
Merge branch 'openmc-dev:develop' into 980_k_search_tally_dervs

pranavkantgaur committed:
remove keff search example dir

pranavkantgaur committed:
move out criticality search tests

pranavkantgaur committed:
remove unused imports

GuySten committed:
simplify code a bit

GuySten committed:
fix docs

GuySten committed:
more simplification

GuySten committed:
fix indentation

GuySten committed:
further simplification

GuySten committed:
further simplification

GuySten committed:
Fix spacing in return statement

GuySten reviewed (approved):
Looks good. 

@pranavkantgaur, If you agree with my changes I will merge tomorrow.

pranavkantgaur committed:
update docstrings and fix a print error

pranavkantgaur commented:
> Looks good.
> 
> @pranavkantgaur, If you agree with my changes I will merge tomorrow.

@GuySten Changes look good, thanks. There was a minor change to handle undefined variable reference for nuclide-density derivatives, I have added that. Also, I have updated the note in the docstrings to not to use this method (yet) for the temperature derivatives.

Besides this, MR description is updated with results from the attached demo script.

paulromano reviewed (changes requested):
Thanks for the PR @pranavkantgaur and the review @GuySten. I'd like to review this before we merge so I'm putting a placeholder review for now.

GuySten committed:
Merge branch 'develop' into 980_k_search_tally_dervs

pranavkantgaur committed:
Merge branch 'openmc-dev:develop' into 980_k_search_tally_dervs

GuySten committed:
Merge branch 'develop' into 980_k_search_tally_dervs
